src/operator/subgraph/mkldnn/mkldnn_fc.cc

```diff
@@ -143,18 +144,34 @@ void SgMKLDNNFCOp::Forward(const OpContext &ctx,
 
     if (mkldnn_param.quantized) {
       CHECK(data.dtype() == mshadow::kInt8 || data.dtype() == mshadow::kUint8);
-      auto data_range = (data.dtype() == mshadow::kInt8) ? kInt8Range : kUint8Range;
-      float data_scale  = data_range / MaxAbs(cached_min_data_, cached_max_data_);
-      float weight_scale = kInt8Range / MaxAbs(cached_min_weight_, cached_max_weight_);
-      float quantized_out_range = IsOutputUint8(full_param_)? kUint8Range : kInt8Range;
-
+      float data_scale  = GetQuantizeScale(data.dtype(), cached_min_data_, cached_max_data_);
+      float weight_scale = GetQuantizeScale(mshadow::kInt8, cached_min_weight_, cached_max_weight_);
       if (has_bias) {
         NDArray bias = in_data[fullc::kBias];
-        float bias_int32_rescale = data_scale * weight_scale *
-            MaxAbs(cached_min_bias_, cached_max_bias_) / kInt8Range;
-
-        cached_bias_ = NDArray(bias.storage_type(), bias.shape(),
-                               bias.ctx(), true, mshadow::kInt32);
+        float bias_scale = GetQuantizeScale(mshadow::kInt8, cached_min_bias_, cached_max_bias_);
+        float bias_int32_rescale = data_scale * weight_scale / bias_scale;
+        // TODO(zhennan): mkldnn has bug to handle INT_MAX in bias, so set the maximum value of bias
+        // to INT_MAX / 2.
+        float bias_max_rescale =
+            MaxValue<int32_t>() / 2 / MaxAbs(cached_min_bias_, cached_max_bias_) / bias_scale;
+        if (bias_int32_rescale > bias_max_rescale) {
+          // avoid overflow on bias
+          bias_int32_rescale = bias_max_rescale;
+          float weight_rescale = bias_int32_rescale * bias_scale / data_scale / weight_scale;
+          cached_weight_.reset(new NDArray(weight.storage_type(), weight.shape(), weight.ctx(),
+                                           true, mshadow::kInt8));
+          int8_t *weight_ptr = weight.data().dptr<int8_t>();
+          int8_t *quantized_weight_ptr = cached_weight_->data().dptr<int8_t>();
+          size_t weight_size = weight.shape().Size();
+#pragma omp parallel for num_threads(engine::OpenMP::Get()->GetRecommendedOMPThreadCount())
+          for (index_t i = 0; i < static_cast<index_t>(weight_size); ++i) {
+            quantized_weight_ptr[i] = weight_ptr[i] * weight_rescale;
```

Comment:
Is it better to consider rounding to nearest than directly truncating?

Reply:
Added.